File: 16-trees2/2-bst-properties/main.py
# BST Property Check - Python Example

class BSTNode:
    def __init__(self, data):
        self.data = data
        self.left = None
        self.right = None

def is_bst(root, min_val=float('-inf'), max_val=float('inf')):
    if root is None:
        return True
    if not (min_val < root.data < max_val):
        return False
    return (is_bst(root.left, min_val, root.data) and
            is_bst(root.right, root.data, max_val))

root = BSTNode(10)
root.left = BSTNode(5)
root.right = BSTNode(15)
root.left.left = BSTNode(3)
root.left.right = BSTNode(7)

print("Is BST:", is_bst(root))

File: 16-trees2/2-bst-properties/main.java
// BST Property Check - Java Example

class BSTNode {
    int data;
    BSTNode left, right;
    BSTNode(int data) {
        this.data = data;
        left = right = null;
    }
}

class BSTProperties {

    public static boolean isBST(BSTNode root, int minVal, int maxVal) {
        if (root == null) return true;
        if (root.data <= minVal || root.data >= maxVal) return false;
        return isBST(root.left, minVal, root.data) && isBST(root.right, root.data, maxVal);
    }

    public static void main(String[] args) {
        BSTNode root = new BSTNode(10);
        root.left = new BSTNode(5);
        root.right = new BSTNode(15);
        root.left.left = new BSTNode(3);
        root.left.right = new BSTNode(7);

        System.out.println("Is BST: " + isBST(root, Integer.MIN_VALUE, Integer.MAX_VALUE));
    }
}
